PSReadLine/Keys.cs

```diff
@@ -268,8 +268,8 @@ void AppendPart(string str)
                     break;
 
                 case '\0':
-                    // This is ugly but familiar.
-                    s = "@";
+                    // This could be a special kind of a modifier key (dead key) on a particular keyboard layout.
+                    s = key.Key.ToString();
```

Comment:
The point in displaying `@` was to be consistent with bash on Linux and to be consistent on Windows when TryGetCharFromConsoleKey failed.

I'm guessing if TryGetCharFromConsoleKey was passed `control: true` that the dead key would have been resolved correctly. If that's the case, maybe the call above on line 229 should retry?

If that doesn't work, can we be smarter here? At least with an English keyboard on Linux, this will typically result in `0`, and on Windows where the `Key` is typically correct, we might see weird things like `D2` or maybe `NumPad2`.

Reply:
> The point in displaying `@` was to be consistent with bash on Linux

Can you please give me an example of bash displaying `@` on Linux?

> If that's the case, maybe the call above on line 229 should retry?

I will check if `TryGetCharFromConsoleKey` can resolve the dead key as expected with `control: true`.

> At least with an English keyboard on Linux, this will typically result in `0`, and on Windows where the `Key` is typically correct, we might see weird things like `D2` or maybe `NumPad2`.

I'm a little confused here ... what do you mean by **this will typically result in 0**? Do you mean the `key.Key` could be `0` with English keyboard on Linux?
And by "**we might see weird things like `D2` or maybe `NumPad2`**", why is that wired? and could you please give me an example? Thanks!

Reply:
> I'm guessing if TryGetCharFromConsoleKey was passed `control: true` ....

You mean when calling `ToUnicode`, set `lpKeyState` as if the `CTRL` key is pressed? This feels weird to me ... and what about on non-Windows? On non-Windows, the code flow will still reach `case '\0'` and `@` will be used ...

Reply:
<kbd>Ctrl+2</kbd> is `0`:

```
jason@jasonsh02:/mnt/c/WINDOWS/system32$ showkey -a

Press any keys - Ctrl-D will terminate this program

^@        0 0000 0x00
```

If you try the same in PowerShell via Console.ReadKey, you'll see that Key is 0.

On Windows, where the Key is captured correctly, <kbd>2</kbd> is `ConsoleKey.D2`.

Reply:
I'm guessing when there is a tty involved, dead keys are not an issue, the CLR and hence PSReadLine won't see control or alt for simple characters.

On Windows, Console.ReadKey is in some ways not helpful reporting keyboard state, although it does allow for some key bindings that aren't supported on Linux.

Reply:
I think your main concern is that with the change in this PR, the same key chord will result in different `PSKeyInfo` instances, which breaks the whole point of abstraction using `PSKeyInfo` (e.g. key bindings).
Let me think about it more. At the meantime, can you please comment on my above question regarding "passing `control:true` to TryGetCharFromConsoleKey"?

Reply:
Updates:
1. `TryGetCharFromConsoleKey` with `control: true` doesn't resolve dead key. For example, with "US international", `'` is a dead key, and `ToUnicode` returns 0 even with `control: true`, meaning the virtual key `Oem7` has no translation for the "US international" keyboard layout.
2. The dead key is not a problem on Linux -- `Console.ReadKey()` doesn't return when only a dead key is pressed. Like you said above, it's likely dead keys are not an issue when there is a tty involved. So this is a windows only problem

In my opinion, the fix should just be not rendering a dead key (a dead key always has `KeyChar` to be `'\0'`). It feels weird to me that we use `'@'` here when `KeyChar` is `'\0'`. Although `showkey -a` prints `^@` for `CTRL+2`, bash doesn't print anything for this key chord, so why does PSReadLine has to print `@` for it?

Reply:
I might be misremembering, but I thought I've seen some Linux terminal displaying `^@` when typing <kbd>Ctrl+2</kbd>, and I emulated it even though it felt weird.

But you are correct - dead keys should not be rendered, so the trick is in distinguishing them, and I don't know without some investigation how you'd do that.

Reply:
Before I go further to find out a way to determine a key chord is a dead key, I would like to understand the point of using `'@'` when `KeyChar` is `'\0'` here.

On Ubuntu 16.04 and 18.04 (maybe other Ubuntu versions too but didn't try), bash doesn't print out anything for <kbd>Ctrl+2</kbd>. Maybe some Linux terminal does display `^@`, but why do we want to emulate that instead of the behavior of Ubuntu bash?
You also mentioned _consistency on windows_, but again, why do we have to print out anything when `KeyChar` is `'\0'` and `TryGetCharFromConsoleKey` fails? Just not rendering the key sounds like a reasonable behavior to me.

Reply:
Also want to add that, in beta.4, <kbd>Ctrl+2</kbd>, <kbd>Ctrl+4</kbd>, <kbd>Ctrl+5</kbd> and etc doesn't result in `^@`, `^\` and `^]` as shown in `showkey -a`, but result in `@`, `\`, `]` because the modifier state is `0`.

Reply:
Consistency is probably impossible - `wsl` doesn't show anything for <kbd>Ctrl+4</kbd> or <kbd>Ctrl+5</kbd>, it's as though you didn't press a key.

Not showing anything is probably fine, but there is an bash key binding [`quoted-insert`](https://www.gnu.org/software/bash/manual/html_node/Commands-For-Text.html#Commands-For-Text) (unimplemented in PSReadLine) that should make it possible to insert control characters like <kbd>Ctrl+2</kbd>.

At any rate - I'm assuming that `key.Key.ToString()` returns `0` on Linux and that's probably not desirable.

Reply:
The `quoted-insert` key binding is interesting. I see <kbd>Ctrl+3</kbd>, <kbd>Ctrl+4</kbd>, <kbd>Ctrl+5</kbd>, and <kbd>Ctrl+6</kbd> print `^[`, `^\`, `^]`, and `^^` as shown in `showkey -a`, but <kbd>Ctrl+2</kbd> still prints nothing. Quite stubborn, right 😄

> At any rate - I'm assuming that key.Key.ToString() returns 0 on Linux and that's probably not desirable.

Agreed. This results in printing `0` for <kbd>Ctrl+2</kbd> on Linux, which is less desired than `@`.
Since the dead key issue is windows-only, I guess we should keep using `@` in case of `KeyChar == '\0'` on non-windows. I will update the code change.